src/HttpClient/Plugin/Authentication/HybridOauth2.php

```diff
@@ -0,0 +1,177 @@
+     *
+     * @var string
+     */
+    public const TOKEN_SCOPES = 'https://www.googleapis.com/auth/cloud-platform';
+
+    /**
+     * Grant type used in an access token request.
+     *
+     * @see https://developers.google.com/identity/protocols/OAuth2ServiceAccount
+     *
+     * @var string
+     */
+    public const GRANT_TYPE = 'urn:ietf:params:oauth:grant-type:jwt-bearer';
+
+    /**
+     * The service account email.
+     *
+     * @var string
+     */
+    protected $email;
+
+    /**
+     * The service account private key.
+     *
+     * @var string
+     */
+    protected $private_key;
+
+    /**
+     * @var \Apigee\Edge\HttpClient\Plugin\Authentication\OauthTokenStorageInterface
+     */
+    protected $tokenStorage;
+
+    /**
+     * @var string
+     */
+    protected $auth_server;
+
+    /**
+     * Hybrid Oauth2 constructor.
+     *
+     * @param string $email
+     *   The service account email.
+     * @param string $private_key
+     *   The service account private key.
+     * @param \Apigee\Edge\HttpClient\Plugin\Authentication\OauthTokenStorageInterface $token_storage
+     *   Storage where access token gets saved.
+     * @param string|null $auth_server
+     *   Authentication server.
+     */
+    public function __construct(string $email, string $private_key, OauthTokenStorageInterface $token_storage, ?string $auth_server = null)
+    {
+        $this->email = $email;
+        $this->private_key = $private_key;
+        $this->tokenStorage = $token_storage;
+        $this->auth_server = $auth_server ?: self::DEFAULT_AUTHORIZATION_SERVER;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function authenticate(RequestInterface $request)
+    {
+        // Get a new access token if token has expired.
+        if ($this->tokenStorage->hasExpired()) {
+            $this->getBearerToken();
+        }
+
+        $accessToken = $this->tokenStorage->getAccessToken();
+
+        if (!empty($accessToken)) {
+            $bearAuth = new Bearer($accessToken);
+            $request = $bearAuth->authenticate($request);
+        }
+
+        return $request;
+    }
+
+    /**
+     * Returns the token storage.
+     *
+     * @return \Apigee\Edge\HttpClient\Plugin\Authentication\OauthTokenStorageInterface
+     */
+    public function getTokenStorage(): OauthTokenStorageInterface
+    {
+        return $this->tokenStorage;
+    }
+
+    /**
+     * Returns a pre-configured client for authorization API calls.
+     *
+     * @return \Apigee\Edge\ClientInterface
+     */
+    protected function authClient(): ClientInterface
+    {
+        return new Client(new NullAuthentication(), $this->auth_server);
+    }
+
+    /**
+     * Retrieves bearer token and saves it to the token storage.
+     *
+     * @psalm-suppress InvalidCatch - Exception by interface can be caught in PHP >= 7.1.
+     */
+    private function getBearerToken(): void
+    {
+        $now = time();
```

Comment:
`time()` shouldn't be used, you can use `\Drupal::time()->getRequestTime();` or an injected service instead, see https://www.drupal.org/node/2785211

Reply:
This is not a Drupal code. I am rather think about whether `time()` or `REQUEST_TIME` is the appropriate value here.

Reply:
doh! you're right, I guess `$_SERVER['REQUEST_TIME']` then unless you really need the exact current time?

Reply:
The docs for the JWT specify that what is needed is `The time the assertion was issued, specified as seconds since 00:00:00 UTC, January 1, 1970.`, so time() should be appropriate here.